libs/langchain_v1/langchain/agents/middleware/pii.py

```diff
@@ -0,0 +1,753 @@
+
+from langchain.agents.middleware.types import AgentMiddleware, AgentState, hook_config
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+    from langgraph.runtime import Runtime
+
+
+class PIIMatch(TypedDict):
+    """Represents a detected PII match in text."""
+
+    type: str
+    """The type of PII detected (e.g., 'email', 'ssn', 'credit_card')."""
+    value: str
+    """The actual matched text."""
+    start: int
+    """Starting position of the match in the text."""
+    end: int
+    """Ending position of the match in the text."""
+
+
+class PIIDetectionError(Exception):
+    """Exception raised when PII is detected and strategy is 'block'."""
+
+    def __init__(self, pii_type: str, matches: list[PIIMatch]) -> None:
+        """Initialize the exception with PII detection information.
+
+        Args:
+            pii_type: The type of PII that was detected.
+            matches: List of PII matches found.
+        """
+        self.pii_type = pii_type
+        self.matches = matches
+        count = len(matches)
+        msg = f"Detected {count} instance(s) of {pii_type} in message content"
+        super().__init__(msg)
+
+
+# ============================================================================
+# PII Detection Functions
+# ============================================================================
+
+
+def _luhn_checksum(card_number: str) -> bool:
+    """Validate credit card number using Luhn algorithm.
+
+    Args:
+        card_number: Credit card number string (digits only).
+
+    Returns:
+        True if the number passes Luhn validation, False otherwise.
+    """
+    digits = [int(d) for d in card_number if d.isdigit()]
+
+    if len(digits) < 13 or len(digits) > 19:
+        return False
+
+    checksum = 0
+    for i, digit in enumerate(reversed(digits)):
+        d = digit
+        if i % 2 == 1:
+            d *= 2
+            if d > 9:
+                d -= 9
+        checksum += d
+
+    return checksum % 10 == 0
+
+
+def detect_email(content: str) -> list[PIIMatch]:
+    """Detect email addresses in content.
+
+    Args:
+        content: Text content to scan.
+
+    Returns:
+        List of detected email matches.
+    """
+    pattern = r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b"
+    matches = []
+
+    for match in re.finditer(pattern, content):
+        matches.append(
+            PIIMatch(
+                type="email",
+                value=match.group(),
+                start=match.start(),
+                end=match.end(),
+            )
+        )
+
+    return matches
+
+
+def detect_ssn(content: str) -> list[PIIMatch]:
+    """Detect US Social Security Numbers in content.
+
+    Detects both formats: 123-45-6789 and 123456789.
+
+    Args:
+        content: Text content to scan.
+
+    Returns:
+        List of detected SSN matches.
+    """
+    matches = []
+
+    # Format with dashes: 123-45-6789
+    pattern_dashed = r"\b\d{3}-\d{2}-\d{4}\b"
```

Comment:
* This will be brittle due to `-` not always being present
* Would prefer to remove also as its too us centric and there are other types of IDs.